### DISCUSSION TITLE
Issue with Duplicate Tailwind CSS Classes in Component Library and Consuming Application

### DISCUSSION THREAD

**Discussion by Kinbaum** [tailwindlabs/tailwindcss#13664] [Help]
I am currently working on a component library (**my-lib**) built with Tailwind v4. The situation involves a consuming application (**my-app**) that pulls components from **my-lib** and imports the compiled CSS from **my-lib** into the root layout.

Both the component library and the consuming application use Tailwind v4. The concern arises from duplicate classes being generated in the final CSS output. For instance, if a button in **my-lib** has the class ﻿`overflow-hidden`, this class is included in the root layout via import.

In **my-app**, if I define `﻿overflow-hidden` on a JSX element, this class gets compiled into the app-specific CSS, resulting in two different ﻿`.overflow-hidden` definitions being shipped to the browser (and understandably so as these are two separate Tailwind processes).

My question is, is there a way to ensure only a single instance of this class is shipped to the browser? Is there something I can do differently to have Tailwind dedupe these classes? Is there a way to deliver the uncompiled CSS from with **my-lib** and then ﻿`@import` it into **my-app**'s CSS file (I tried the second option using a monorepo, but I was getting a bunch of post-css errors)?

**Comment by wongjn** [comment]
Do you have any custom CSS in the **input** CSS of `my-lib`? If not, you could try letting `my-app` compile all the Tailwind CSS, for itself and `my-lib`. If you **do** have some **input** CSS in `my-lib`, you could try to import the source CSS into `my-app`'s **input** CSS.

**Reply by Kinbaum** [reply]
Yeah my-lib has a combination of default utils and custom classes that are generated using the new @theme syntax. The idea is that my-app would simply override the css variables these classes point to for custom theming. This works but only with the compiled version (that's how I noticed duplicates). I tried importing the uncompiled styles directly into my-app and post-css errored out on the path. I'm not using a tailwind config as this is v4.

**Reply by Kinbaum** [reply]
For context, I'm using a monorepo (turborepo) to build the lib / docs. And I'm just attempting to do a workspace import. Even though I specify my lib exports the uncompiled css, it doesn't seem to be able to locate it. Also, using a relative import (../../) doesn't work either. Postcss wont error, I just don't get any styling.

**Reply by Kinbaum** [reply]
I think I can see what's happening when I import the uncompiled CSS from my-lib into my-app. 
Any custom classes / colors that I set, those are being imported.

However, when I have a color defined in my-lib `@theme` like `--background-color-button-solid`, I can see the custom CSS property in the final output CSS in my-app, but the corresponding class name on my button component `bg-button-solid` does not get created. Perhaps it's improperly purged? @adamwathan any thoughts in this?

**Comment by tounsoo** [comment]
Hi @Kinbaum, I'm also thinking of creating a test ui-lib using tailwindcss4, how has it been working out and is there a repo that I can take a peak? 😉

**Comment by Kinbaum** [comment]
@adamwathan @wongjn In order to import the uncompiled css from my-lib into my-app, I had to modify the postcss config to the following

```js
const path = require('node:path');

module.exports = {
	plugins: {
		'@tailwindcss/postcss': {
			base: path.resolve(__dirname, '../../packages/registry')
		}
	}
};
```

While I can now see my library styles coming through, the application-specific styles are no longer rendered due to the base directory changing. If I use the root of my monorepo I get an infinite loop it seems as there are two separate tailwind instances here. Outside of my monorepo, I want to distribute this lib as an npm package, so not sure if it'll work properly. 

The more I mess with this, I'm wondering if node_modules is just being ignored completely?
I'm using pnpm and I can see my lib being symlinked in the node_modules of my app. Is there way to allow a specifc package from node_modules to be included in the glob?
